plugin/pathvirt/pathvirt.cpp

```diff
@@ -0,0 +1,250 @@
+    }
+
+    /* yes, should swap */
+
+    /* check if path is in list of registered paths to swap out */
+    int index = clfind(oldPathPrefixList, path, &oldPathPtr);
+    if (index == -1)
+        return false;
+
+    /* found it in old list, now get a pointer to the new prefix to swap in*/
+    char *newPathPtr = clget(newPathPrefixList, index);
+    if (newPathPtr == NULL)
+        return false;
+
+    size_t newElementSz = clgetsize_ptr(newPathPrefixList, newPathPtr);
+    size_t oldElementSz = clgetsize_ptr(oldPathPrefixList, oldPathPtr);
+
+    /* temporarily null terminate new element */
+    newPathPtr[newElementSz] = '\0';
+
+    /* finally, create full path with the new prefix swapped in */
+    newPath = newPathPtr;
+    newPath += "/";
+    newPath += (path + oldElementSz);
+
+    /* repair the colon list */
+    newPathPtr[newElementSz] = ':';
+
+    return true;
+}
+
+/*
+ * DMTCP Setup
+ */
+
+void
+dmtcp_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
+{
+    /* NOTE:  See warning in plugin/README about calls to printf here. */
+    switch (event) {
+    case DMTCP_EVENT_INIT:
+    {
+        /* On init, check if they've specified paths to virtualize via
+           DMTCP_PATH_PREFIX env */
+        char *oldEnv = getenv(ENV_DPP);
+        if (oldEnv) {
+            /* if so, save it to buffer */
+            snprintf(oldPathPrefixList, sizeof(oldPathPrefixList), "%s",
+                     oldEnv);
+        }
+        break;
+    }
+    case DMTCP_EVENT_RESTART:
+    {
+        /* necessary since we don't know how many bytes dmtcp_get_restart_env
+           will write */
+        memset(newPathPrefixList, 0, sizeof(newPathPrefixList));
+
+        /* Try to get the value of ENV_DPP from new environment variables,
+         * passed in on restart */
+        int ret = dmtcp_get_restart_env(ENV_DPP, newPathPrefixList,
+                                        sizeof(newPathPrefixList) - 1);
+
+        /* ret == -1 is fine everything else is not */
+        if (ret < -1) {
+            /* this will be run into first, since dmtcp_get_restart_env's
+               MAXSIZE is currently less than MAX_ENV_VAR_SIZE */
+            JASSERT(ret != -3).Text("dmtcpplugin: DMTCP_PATH_PREFIX exceeds "
+                    "dmtcp_get_restart_env()'s MAXSIZE. Use a shorter "
+                    "environment variable or increase MAXSIZE and recompile.");
+
+            /* if MAXSIZE is resized in the future to be greater than
+               MAX_ENV_VAR_SIZE (to get past the first assert) we will then
+               trigger this assert */
+            JASSERT(ret != -2).Text("pathvirt: DMTCP_PATH_PREFIX exceeds "
+                    "maximum size (10kb). Use a shorter environment variable "
+                    "or increase MAX_ENV_VAR_SIZE and recompile.");
+
+            /* all other errors */
+            JASSERT(ret >= 0).Text("Fatal error retrieving DMTCP_PATH_PREFIX "
+                    "environment variable.");
+        }
```

Comment:
@gc00 : I'd appreciate a review of this code. Previously there was no handling for dmtcp_get_restart_env() failures, but this should be fairly robust.